### PR TITLE
Add config option to specify json float serialization precision

### LINKED ISSUE #875: Maximum precision for float serialization
We should explicitly set the number of decimal places to use for float serialization 

- `queue_pop_time_delta`

### PR DESCRIPTION
**Pull request by Gatsik**
Closes #875

### CODE REVIEW

**Comment by Gatsik** [comment]
`test_ladder_game_draw_bug ` and `test_game_ended_broadcasts_rating_update` are failing probably because of customized precision
Do you want this custom precision only for `queue_pop_time_delta` field in `matchamker_info` message?

**Comment by Askaholic** [comment]
Hey, nice work! I think it makes sense to use the same precision for everything, so those tests probably need to be adjusted to have higher tolerance for error.

The second one is interesting because it means there will be some edge cases where the rating change is so small that it doesn’t appear to change at all. I think that’s ok though since the full change history will still be available through the replay details page and the rating graph.

**Comment by Askaholic** [comment]
One thing we have to keep in mind is that if we’re messing with the json encoding, we need to be very careful of the performance impact since this is one of the hottest functions in the entire code base. It would be amazing if we could just override the float serialization code and leave the rest of the implementation untouched. I found a little snippet that does this although it uses the `mock.patch` functionality to override private functions from the Json library. Still, this may be a good source of inspiration: 

https://gist.github.com/Sukonnik-Illia/ed9b2bec1821cad437d1b8adb17406a3

**Comment by Gatsik** [comment]
It looks like we can't do much without creating custom python package that will use compiled C code.
I did some comparison between different methods using slightly modified script taken from https://developpaper.com/speed-comparison-of-five-json-libraries-in-python/ and here are the results of encoding on my machine:
```Console
patching (example given above): 107.800
current (this pr's code):  28.629

simplejson: 6.981
ujson: 3.503
stdlib json: 4.500
```
Round floats approach (https://stackoverflow.com/a/53798633):
```Console
simplejson: 14.196
ujson: 10.481
stdlib json: 11.109
```
Overriding `iterencode` function, which contains `floatstr` method, directly without patching:
```Console
15.991
```
Compiling custom mix of `json` and `simplejson` (most of the code is a copy from https://github.com/python/cpython/pull/13233/, built with `setup.py` from `simplejson`):
```Console
6.589
```
Time is in seconds, the data for encoding is: 
```Python
{
    "command": "game_info",
    "visibility": "public",
    "password_protected": True,
    "uid": 13,
    "title": "someone's game",
    "state": "playing",
    "game_type": "custom",
    "featured_mod": "faf",
    "sim_mods": {},
    "mapname": "scmp_009",
    "map_file_path": "maps/scmp_009.zip",
    "host": "Foo",
    "num_players": 2,
    "launched_at": 1111111111.1112312312312,
    "rating_type": "faf",
    "rating_min": None,
    "rating_max": None,
    "enforce_rating_range": False,
    "team_ids": [
        {
            "team_id": 1,
            "player_ids": [1],
        },
        {
            "team_id": 2,
            "player_ids": [2],
        },
    ],
    "teams": {
        1: ["Foo"],
        2: ["Bar"],
    },
}
```
There is also a package called `orjson` which is super fast, but requires `dict` keys to be strings, so with modified initial data, in which all keys are strings, the result is:
```Console
0.624
```
Otherwise, using "round floats" approach in which we also convert `dict` keys to `str`:
```Console
8.457
```

File: tests/unit_tests/test_protocol.py [hunk 1]
```diff
@@ -256,3 +259,117 @@ async def test_read_when_disconnected(protocol):
...
+        ],
+        "teams": {
+            1: ["Foo"],
+            2: ["Bar"],
+        },
+    }
+    assert new_encode(game_info_dict) == old_encode(game_info_dict)
+
+
+def test_json_encoder_encodes_dicts():
```
Comment (Askaholic) [inline tests/unit_tests/test_protocol.py:352]: This could be a good test to write using `hypothesis`. You would just need to make a composite strategy that can generate any of the type's you're testing (bools, None, strings, ints, floats). Hypothesis will take care of figuring out the actual data.

File: tests/unit_tests/test_protocol.py [hunk 2]
```diff
@@ -256,3 +259,117 @@ async def test_read_when_disconnected(protocol):
...
+        json_encoder.encode({wrong_dict_key: 'a'})
+
+
+def test_json_encoder_encodes_lists_and_tuples():
+    old_encode = json.JSONEncoder(separators=(",", ":")).encode
+    new_encode = json_encoder.encode
+
+    weird_list = [
+        True, None, '"double_quoted"', "'single_quoted'", 1, 1.22,
+        {'a': (1, 2, 3)}, (1, 2, 3), ({'a': 1}, True, 12.33),
```
Comment (Askaholic) [inline tests/unit_tests/test_protocol.py:372]:
```suggestion
        {"a": (1, 2, 3)}, (1, 2, 3), ({"a": 1}, True, 12.33),
```

File: server/protocol/protocol.py [hunk 3]
```diff
@@ -4,10 +4,26 @@
...
 import server.metrics as metrics
+from server.config import config
 
 from ..asyncio_extensions import synchronizedmethod
 
-json_encoder = json.JSONEncoder(separators=(",", ":"))
+
+class CustomJSONEncoder(json.JSONEncoder):
+    # taken from https://stackoverflow.com/a/53798633
+    def encode(self, o):
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/protocol/protocol.py:14]: Oh sorry, I see this is the 'round floats' approach you mentioned in your other comment.

File: server/protocol/protocol.py [hunk 4]
```diff
@@ -4,10 +4,26 @@
...
+        def round_floats(o):
+            if isinstance(o, float):
+                return round(o, config.JSON_FLOAT_MAX_DIGITS)
+            if isinstance(o, dict):
+                return {k: round_floats(v) for k, v in o.items()}
+            if isinstance(o, (list, tuple)):
+                return [round_floats(x) for x in o]
+            return o
+        return super().encode(round_floats(o))
```
Comment (Askaholic) [inline server/protocol/protocol.py:23]: Maybe we should add a config option to disable this feature just in case it turns out to be too costly in production. Although, I think even if it doubles the json encoding time that will still be fine judging by the profiling information I've collected from the server in the past.

File: server/servercontext.py [hunk 5]
```diff
@@ -132,8 +132,8 @@ async def client_connected(self, stream_reader, stream_writer):
             asyncio.CancelledError,
         ):
             pass
-        except Exception:
-            self._logger.exception()
+        except Exception as e:
+            self._logger.exception(e)
```
[the remark this answers is not in the document]
Comment (Gatsik) [reply server/servercontext.py:136]: Because `logger.exception` has 1 required positional argument (https://docs.python.org/3.8/library/logging.html#logging.Logger.exception) and without it will throw `TypeError`

File: server/protocol/protocol.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gatsik) [reply server/protocol/protocol.py:14]: I don't quite understand what do you mean by a wrapper, because if it requires us to call this wrapper every time we do some math, then why don't just use `round`?
Rewriting representation of the float won't help, the C code will operate with value and encode all the digits

File: server/protocol/protocol.py (hunk 4 above)
Comment (Gatsik) [reply server/protocol/protocol.py:23]: Yes, sure, what would you call it?

File: server/protocol/protocol.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/protocol/protocol.py:14]: Something similar to this, but using our own class instead of Decimal: https://stackoverflow.com/a/3885198

```python3
class PFloat:
    def __init__(self, value: float, precision: int):
        self.value = value
        self.precision = precision
```

And then in the `to_dict` method you'd have to wrap the floats in this class:

```python3
def to_dict(self):
    return {
        "some_float": PFloat(self.some_float, 2)
    }
```

File: server/servercontext.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/servercontext.py:136]: Hah, good catch!

File: server/protocol/protocol.py (hunk 4 above)
Comment (Askaholic) [reply server/protocol/protocol.py:23]: I think `JSON_ROUND_FLOATS` would be good. And then rename the other one to `JSON_ROUND_FLOATS_MAX_DIGITS` or `JSON_ROUND_FLOATS_PRECISION`

File: server/protocol/protocol.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gatsik) [reply server/protocol/protocol.py:14]: Yeah, I can make a class, but I think it is unnecessary complication, because we already have built-in `round` function which effectively does the same thing

[the remark this answers is not in the document]
Comment (Gatsik) [reply server/protocol/protocol.py:14]: Like, we have 2 options:
  1. Rewrite json's `encode` method to change encoding of all floats
  2. Rewrite every `to_dict` method where we will round every float with its own precision (the "more control" mentioned above)

File: server/protocol/protocol.py (hunk 4 above)
Comment (Gatsik) [reply server/protocol/protocol.py:23]: Should mocking be also conditional in tests?

Comment (Askaholic) [reply server/protocol/protocol.py:23]: I don't think it's too important to have tests for the config values. They're generally simple enough that it doesn't matter

**Comment by Gatsik** [comment]
I don't think I will come back to look at this
